src/cdk/object.ts

```diff
@@ -0,0 +1,111 @@
+import { MiraConfig, Account } from '../config/mira-config'
+import MiraEnv from './env'
+
+/**
+ * @class Responsible for naming resources and establishing asynchronous
+ * object initialization pattern.
+ */
+export class MiraObject {
+    /* eslint-disable-next-line */
+    initialized: Promise<any>
+    name: string
+    resourceType: string
+    constructor (name: string, resourceType = 'resource') {
+      if (resourceType === 'resource') {
+        console.warn('Warning: resource type not defined for ' +
+              `${this.constructor.name}, defaulting to 'resource' as name`)
+      }
+      this.resourceType = resourceType
+      this.name = name
+      /* eslint-disable-next-line */
+      this.initialized = new Promise(async (resolve) => {
+        try {
+          // TODO: With stack, we won't run this code.
+          await this.preInitialize()
+          await this.initialize()
+          // TODO: With stack, we won't run this code.
+          await this.postInitialize()
+          resolve()
+        } catch (e) {
+          console.warn(`Initialization of ${this.constructor.name} object failed:`, e)
+          throw new Error(e)
```

Comment:
This will result in an unhandled promise rejection warning when the error is thrown and in the next version of Node that will crash the application.

Could something like this work?:
```
this.initialized = this.preInitialize()
      .then(() => this.initialize())
      .then(() => this.postInitialize())
      .then(() => true)
      .catch(e => {
        console.warn(`Initialization of ${this.constructor.name} object failed:`, e)
        throw e
      })
```

Reply:
I think the intent roughly is to crash the program.  The stack.ts file or the app.ts file is supposed to catch these errors and exit gracefully without building a Cfn template file.